### PR TITLE
[AArch64] Don't emit Neon in streaming[-compatible] functions with -fzero-call-used-regs

### PR DESCRIPTION
**Pull request by MacDue**
Previously, with `-fzero-call-used-regs` clang/LLVM would incorrectly emit Neon instructions in streaming functions, and streaming-compatible functions without SVE.

With this change:

* In streaming functions, Z/p registers will be zeroed
* In streaming compatible functions w/o SVE, D registers will be zeroed
  - (As Neon vector instructions are illegal including `movi v..`)

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp [hunk 1]
```diff
@@ -9694,19 +9694,26 @@ void AArch64InstrInfo::buildClearRegister(Register Reg, MachineBasicBlock &MBB,
...
       .addImm(0)
       .addImm(0);
-  } else {
+  } else if (STI.isNeonAvailable()) {
     BuildMI(MBB, Iter, DL, get(AArch64::MOVIv2d_ns), Reg)
       .addImm(0);
+  } else {
+    // This is a streaming-compatible function without SVE. We don't have full
+    // Neon (just FPRs), so we can at most use the first 64-bit sub-register.
+    // So given `movi v..` would be illegal use `fmov d..` instead.
```
Comment (MacDue) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:9713]: I'm pretty sure this is correct. I don't think there's a way to zero the full 128 bits without Neon or SVE in a streaming-compatible function (we at most have the D registers).

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp [hunk 2]
```diff
@@ -9694,19 +9694,26 @@ void AArch64InstrInfo::buildClearRegister(Register Reg, MachineBasicBlock &MBB,
                                           MachineBasicBlock::iterator Iter,
                                           DebugLoc &DL,
                                           bool AllowSideEffects) const {
-  const MachineFunction &MF = *MBB.getParent();
+  MachineFunction &MF = *MBB.getParent();
```
Comment (sdesmalen-arm) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:9697]: Is this change necessary?

Comment (MacDue) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.cpp]: No (I used the wrong function at first which was non-const)

**Review by sdesmalen-arm — APPROVED** [review]
